silhouette/app/com/mohiva/play/silhouette/api/services/AuthenticatorService.scala

```diff
@@ -20,9 +20,52 @@
 package com.mohiva.play.silhouette.api.services
 
 import com.mohiva.play.silhouette.api.{ Authenticator, LoginInfo }
+import play.api.libs.concurrent.Execution.Implicits._
 import play.api.mvc.{ Headers, RequestHeader, Result }
 
 import scala.concurrent.Future
+import scala.language.implicitConversions
+import scala.util.Try
+
+/**
+ * A marker result which indicates that an operation on an authenticator was processed and
+ * therefore it shouldn't updated automatically.
+ *
+ * Due the fact that the update method gets called on every subsequent request to update the
+ * authenticator related data in the backing store and in the result, it isn't possible to
+ * discard or renew the authenticator simultaneously. This is because the "update" method would
+ * override the result created by the "renew" or "discard" method, because it will be executed
+ * as last in the chain.
+ *
+ * As example:
+ * If we discard the session in a Silhouette action then it will be removed from session. But
+ * at the end the update method will embed the session again, because it gets called with the
+ * result of the action.
+ *
+ * @param result The result to wrap.
+ */
+class AuthenticatorResult(result: Result) extends Result(result.header, result.body, result.connection)
+
+/**
+ * The companion object.
+ */
+object AuthenticatorResult {
+
+  /**
+   * Converts a [[Result]] into [[AuthenticatorResult]]
+   */
+  implicit def toAuthenticatorResult(result: Result): AuthenticatorResult = apply(result)
+  implicit def toAuthenticatorResultF(result: Future[Result]): Future[AuthenticatorResult] = result.map(apply)
+  implicit def toAuthenticatorResultT(result: Try[Result]): Try[AuthenticatorResult] = result.map(apply)
+
```

Comment:
Background:

When a user logs in, silhouette embeds an Authenticator in the http response.  The client (ie. the browser) has to repeat that authenticator in subsequent requests.  To limit the lifetime of the authenticator, silhouette keeps an idleTimeout in authenticator. That idleTimeout gets refreshed for every request to keep an active session from timing out. To make this process transparent to developer code, Siljouette's ActionBuilders handle this automatically: they extract the authenticator from the request, update the idleTimeout and embed the updated authenticator in the response.





The problem occurs when a controller wants to embed a different authenticator in the response.  This PR tries to solve that problem by introducing a wrapper type for the result that will signal to Silhouette's ActionBuilders to not overwrite the newly embedded authenticator.  In this case the marker is called AuthenticatorResult.




Hope that helps

On Fri, Apr 3, 2015 at 5:21 PM, Nadir Muzaffar <notifications@github.com>
wrote:

>> + * @param result The result to wrap.
>> + */
>> +class AuthenticatorResult(result: Result) extends Result(result.header, result.body, result.connection)
>> +
>> +/**
>> + * The companion object.
>> + */
>> +object AuthenticatorResult {
>> +
>> +  /**
>> +   * Converts a [[Result]] into [[AuthenticatorResult]]
>> +   */
>> +  implicit def toAuthenticatorResult(result: Result): AuthenticatorResult = apply(result)
>> +  implicit def toAuthenticatorResultF(result: Future[Result]): Future[AuthenticatorResult] = result.map(apply)
>> +  implicit def toAuthenticatorResultT(result: Try[Result]): Try[AuthenticatorResult] = result.map(apply)
>> +
> Sorry, I've trying to follow you guys. Could anyone clarify what "marker" refers to?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/mohiva/play-silhouette/pull/323/files#r27756751